File: leetcode/67.add-binary.py
# @leet start
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        res: list[str] = []
        carry = 0

        for i in range(max(len(a), len(b))):
            dig_sum = carry

            if i < len(a):
                dig_sum += int(a[-1 - i])

            if i < len(b):
                dig_sum += int(b[-1 - i])

            res.append(str(dig_sum % 2))
            carry = dig_sum // 2

        if carry:
            res.append("1")

        return "".join(res[::-1])


# @leet end

File: leetcode/67.add-binary.ts
// @leet start
function addBinary(a: string, b: string): string {
  const resultReversed: string[] = [];
  let carry = 0;

  for (let i = a.length - 1, j = b.length - 1; i >= 0 || j >= 0; i--, j--) {
    const val =
      carry +
      (i >= 0 && a[i] === "1" ? 1 : 0) +
      (j >= 0 && b[j] === "1" ? 1 : 0);
    resultReversed.push(String(val % 2));
    carry = Math.floor(val / 2);
  }

  if (carry) {
    resultReversed.push("1");
  }

  return resultReversed.reverse().join("");
}
// @leet end
